src/values/AbstractValue.js

```diff
@@ -117,68 +169,34 @@ export default class AbstractValue extends Value {
     );
   }
 
-  getHash(): number {
-    return this.hashValue;
-  }
-
   getBuildNode(): AbstractValueBuildNodeFunction | BabelNodeExpression {
     invariant(this._buildNode);
     return this._buildNode;
   }
 
-  buildNode(args: Array<BabelNodeExpression>): BabelNodeExpression {
-    let buildNode = this.getBuildNode();
-    return buildNode instanceof Function
-      ? ((buildNode: any): AbstractValueBuildNodeFunction)(args)
-      : ((buildNode: any): BabelNodeExpression);
+  getHash(): number {
+    return this.hashValue;
   }
 
-  hasIdentifier() {
-    return this._buildNode && this._buildNode.type === "Identifier";
+  getType() {
+    return this.types.getType();
   }
 
   getIdentifier() {
     invariant(this.hasIdentifier());
     return ((this._buildNode: any): BabelNodeIdentifier);
   }
 
-  addSourceLocationsTo(locations: Array<BabelNodeSourceLocation>) {
-    if (this._buildNode && !(this._buildNode instanceof Function)) {
-      if (this._buildNode.loc) locations.push(this._buildNode.loc);
-    }
-    for (let val of this.args) {
-      if (val instanceof AbstractValue) val.addSourceLocationsTo(locations);
-    }
+  hasIdentifier() {
+    return this._buildNode && this._buildNode.type === "Identifier";
   }
 
-  addSourceNamesTo(names: Array<string>) {
-    let gen = this.$Realm.preludeGenerator;
-    function add_intrinsic(name: string) {
-      if (name.startsWith("_$")) {
-        if (gen === undefined) return;
-        add_args(gen.derivedIds.get(name));
-      } else if (names.indexOf(name) < 0) {
-        names.push(name);
-      }
-    }
-    function add_args(args: void | Array<Value>) {
-      if (args === undefined) return;
-      for (let val of args) {
-        if (val.intrinsicName) {
-          add_intrinsic(val.intrinsicName);
-        } else if (val instanceof AbstractValue) {
-          val.addSourceNamesTo(names);
-        } else if (val instanceof StringValue) {
-          if (val.value.startsWith("__")) {
-            names.push(val.value.slice(2));
-          }
-        }
-      }
-    }
-    if (this.intrinsicName) {
-      add_intrinsic(this.intrinsicName);
-    }
-    add_args(this.args);
+  implies(val: AbstractValue): boolean {
+    // Neither this nor val is a known value, so we need to some reasoning based on the structure
```

Comment:
(The intention is that) this works on any kind of AbstractValue, not just those which are of type BooleanValue, right?

Reply:
Since any value can be converted to a Boolean and many path conditions will be "thruthy" rather than Boolean valued, yes.